src/process.rs

```diff
@@ -330,27 +330,65 @@ fn read_process_pacs_from_iter<I>(
     iter: I,
     process_ids: &HashSet<Rc<str>>,
     commodities: &HashMap<Rc<str>, Rc<Commodity>>,
+    flows: &HashMap<Rc<str>, Vec<ProcessFlow>>,
 ) -> Result<HashMap<Rc<str>, Vec<Rc<Commodity>>>>
 where
     I: Iterator<Item = ProcessPAC>,
 {
     // Keep track of previous PACs so we can check for duplicates
-    let mut pacs = HashSet::new();
-
-    iter.map(|pac| {
-        let process_id = process_ids.get_id(&pac.process_id)?;
-        let commodity = commodities.get(pac.commodity_id.as_str());
-
-        match commodity {
-            None => bail!("{} is not a valid commodity ID", &pac.commodity_id),
-            Some(commodity) => {
-                ensure!(pacs.insert(pac), "Duplicate PACs found");
-
-                Ok((process_id, Rc::clone(commodity)))
+    let mut existing_pacs = HashSet::new();
+
+    // Build hashmap of process ID to PAC commodities
+    let pacs = iter
+        .map(|pac| {
+            let process_id = process_ids.get_id(&pac.process_id)?;
+            let commodity = commodities.get(pac.commodity_id.as_str());
+
+            // Check that commodity is valid and PAC is not a duplicate
+            match commodity {
+                None => bail!("{} is not a valid commodity ID", &pac.commodity_id),
+                Some(commodity) => {
+                    ensure!(existing_pacs.insert(pac), "Duplicate PACs found");
+                    Ok((process_id, Rc::clone(commodity)))
+                }
             }
+        })
+        .process_results(|iter| iter.into_group_map())?;
+
+    // Check that PACs for each process are either all inputs or all outputs
+    for (process_id, pacs) in pacs.iter() {
+        let flows = flows.get(process_id).unwrap();
```

Comment:
I'd say why the `unwrap()` is safe here:

```suggestion
        // This is safe as every process has associated flows
        let flows = flows.get(process_id).unwrap();
```